### PR TITLE
[WIP] Automated Testing

### PR DESCRIPTION
**Pull request by bryanchriswhite**
# Current Plan

### The Directory Structure:

There is a "test binary" at `bin/test.js`. This "binary acts solely as an entry point to load babel so that the tests can use our usual transforms/syntax. This binary is what's executed by `npm test`.

The test binary executes the `scripts/test/index.js`. This script uses [`commander`](https://github.com/tj/commander.js) to parse arguments (more on that later) and kicks off the various types of tests based on those args.

#### Unit

Unit tests should live in the same directory, next to the unit they're testing with a `-spec` suffix to the filename.

#### E2E

E2E tests don't necessarily (in fact, rarely) correlate to the directory structure of the units. As a result, they should be organized by functionality or features. The example below doesn't contain any such organization yet due to the in-progress nature of the work. E2E tests have a `-e2e` suffix in their filename to help distinguish them.

#### Visual Regression

Similar to E2E, visual regression tests have their own logical organization structure which may actually correlate reasonably well with the E2E tests (time will tell). Again, here the example is incomplete due to the in-progress nature of the work but at this point. 

I believe a good starting point for this structure would be to make directories out of the filenames (as compared to the e2e dir structure, minus the `-e2e` suffix), those directories then contain an `index.js` which will contain the test code and one or more screenshot that the test will use for comparison. I'm thinking it might be reasonable to make the filename of the screenshot the commit hash of the last commit or something like that. Visual regression tests have a `-viz` suffix in their filename.

```
bin
└── test.js
scripts
└── test
    ├── e2e.js
    ├── index.js
    ├── unit.js
    └── visual.js
src
└── components
     └── Auth
        └── SignupForm
            ├── index-spec.js
            └── index.js
e2e
└── signup-e2e.js
visualRegression
└── signup
     ├── index-viz.js
     └── <commit-hash>.png
```

### Types of Tests & Tooling:

All types of tests rely on the [mocha](https://mochajs.org/) testing framework and the [chai](http://chaijs.com/) assertion libraries (typically `expect`).

+ `unit`: Tests which isolate the smallest possible units and mock/stub everything else. Unit tests will be run in a browser JS engine and DOM (chrome/firefox/phantomJS/slimerJS) using the [karma](https://karma-runner.github.io/1.0/index.html) test runner, and [enzyme](http://airbnb.io/enzyme/index.html) for react-specific test tooling.
+ `e2e` (end-to-end): Tests which start the application and exercise from it's public interface and ensure the appropriate side effects are perceived (http/api response, DB records updated, external http api called, etc). E2e tests will automate a real browser (chrome/firefox/phantomJS/slimerJS) using [nightwatch](http://nightwatchjs.org/) (or alternatively [casperjs](https://github.com/casperjs/casperjs)).
+ `visual` (visual regression): Tests which compare screenshots of the running application pages and compare them to screenshots from previous versions. Visual regression will be tested with [phantomcss](https://github.com/Huddle/PhantomCSS) which relies on [casperjs](https://github.com/casperjs/casperjs).

### Running The Tests:

You can run all the tests with a simple `npm test` but if you want to get more specific you can pass some arguments to narrow your selection.

**NOTE: to pass arguments "through `npm`" you have to put a `--` inbetween `npm test` and your args e.g.: (`npm test -- -u` or `npm test -- --unit`)**

The CLI arguments/options that are available are defined in [scripts/test/index.js](https://github.com/bryanchriswhite/bridge-gui/blob/testing-63/scripts/test/index.js#L12).

You can print the usage with `npm test -- -h`:

```
  Usage: test [options]

  Options:

    -h, --help     output usage information
    -V, --version  output the version number
    -a, --all      Run all tests
    -u, --unit     Run unit tests
    -e, --e2e      Run end-to-end tests
    -v, --visual   Run visual regression tests
```

_Note on commander: commander is a sexy option parser that lets you pass args similar to how you would expect a unix binary to (in addition to many other cool features). The following are all equivalent:_

+ `npm test -- -u -e`
+ `npm test -- -e -u`
+ `npm test -- -ue`
+ `npm test -- -eu`
+ `npm test -- --unit -e`
+ `npm test -- -e --unit`
+ `npm test -- -u --e2e`
+ `npm test -- --unit --e2e`

### CODE REVIEW

**Comment by bryanchriswhite** [comment]
### NOTE: this PR is upstream of #64 - Babel 6 Upgrade

**Comment by bookchin** [comment]
This all sounds great to me. Party on. :tada:

File: package.json [hunk 1]
```diff
@@ -100,6 +104,7 @@
     "karma-sourcemap-loader": "^0.3.5",
     "karma-webpack": "^1.7.0",
     "mocha": "^2.3.3",
+    "nightwatch": "^0.9.5",
     "phantomjs": "^1.9.18",
     "phantomjs-polyfill": "0.0.1",
     "react-hot-loader": "^1.3.0",
```
Comment (bryanchriswhite) [inline package.json]: _NOTE: npm likes things in alphabetical order hence, the exaggerated diff here_

Packages modified (most of these are already meged in #64):
+ +`nightwatch`
+ +`glob`
+ +`commander`
+ +`colors`
+ +`babel-preset-react`
+ +`babel-preset-stage-0`
+ +`babel-preset-es2015`
+ +`babel-plugin-transform-decrators-legacy`
+ -`babel` (not needed in babel 6)
+ -`babel-runtime` (not needed in babel 6)
+ -`babel-plugin-react-transform` (replaced by `babel-preset-react`)

File: scripts/test/webpack/e2e.config.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  //   fs: 'empty',
+  //   net: 'empty',
+  //   tls: 'empty'
+  // },
+  // plugins: [
+  //   optimizations
+  //   new webpack.optimize.DedupePlugin(),
+  //   new webpack.optimize.OccurenceOrderPlugin(),
+  // ]
+};
```
Comment (bryanchriswhite) [inline scripts/test/webpack/e2e.config.js]: This file isn't being used yet but would be if we decide to use webpack to prebuild and cache the test suite

File: scripts/test/webpack/e2e.config.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  module: {
+    noParse: [],
+    loaders: [
+      {test: /\.js$/, exclude: /node_modules/, loaders: ['babel']},
+      {test: /\.json$/, loader: 'json'}
+    ]
+  },
+  progress: false,
+  target: 'node'
+};
```
Comment (bryanchriswhite) [inline scripts/test/webpack/e2e.config.js]: This file isn't being used currently but would be if we use webpack for building/watching/caching the ES6 test suite

**Comment by phutchins** [comment]
Tests that should pass are passing and tests run as expected.

**Pull request merged**